### DISCUSSION TITLE
Provide a lightweight struct package for ViewModel.

### DISCUSSION THREAD

**Discussion by 1amageek** [firebase/firebase-ios-sdk#8595] [Ideas]
The iOS development process has changed dramatically with the advent of SwiftUI. The biggest advantage is that you can now quickly check and change the View by hot reloading. However, the Firebase SDK is very heavy and takes a long time to build.

My solution to this is to create only the View in a sub-project that eliminates the Firebase SDK dependency. Then I use SPM to integrate it into the main project.
This allows us to separate it from the Firebase SDK and build it smoothly, but it still doesn't solve all the problems.

Firestore has specific classes such as DocumentReference, Timestamp, FieldValue, etc. We can't declare them in the subproject to remove the dependency on Firebase SDK.

In many cases, there are cases where the ViewModel contains ID,Timestamp. However, since FirebaseFirestoreSwift uses PropertyWrapper, it is also not possible to define it as a Protocol well.

## Solution

I believe the solution is to define lightweight DocumentReference, Timestamp and FieldValue structs in Firebase SDK iOS.

The DocumentReference and Timestamp defined in the ViewModel are mainly used for display, and there is no need to control the DB controls.

I think the `Swift4ServerTimestamp` approach is a solution that makes sense. I think the `Swift4ServerTimestamp` approach is a reasonable solution because it is very lightweight and can do its job well.

**Comment by maksymmalyhin** [comment]
@1amageek Thank you for the feedback, detailed explanation and the proposed solution! 

I'm not sure I completely understand all the details, so I would like to as a couple questions:

1. When mentioning hot reload are you referring to SwiftUI previews? I'm not aware of natively supported hot reload of an app, so I assume it's about previews, but I would like to double check in case I'm missing something
2. My naive assumption is that for SwiftUI preview to be rendered faster you may use a preview specific ViewModel that doesn't have extra heavy dependencies like Firestore SDK. To make it working smoothly the View should depend on a protocol defining the ViewModel API, so any ViewModel implementation can be used. Also it requires the ViewModel API being implementation agnostic, e.g. no Firebase classes should be used in the API to make a lightweight preview ViewModel implementation possible. As far as I understand you encounter some inconvenience with this approach. Could you please provide a bit more details on it so we have better understanding of the issue.
3. A sample project illustrating the inconvenience you encounter will help to build the shared understanding. Could you please post such a sample project?

**Reply by 1amageek** [reply]
Thanks for the question.
I used the hot reload expression, but you are right, it is a function of the preview.

Prepare a simple project.

**Comment by 1amageek** [comment]
<img width="785" alt="スクリーンショット 2021-09-01 14 24 35" src="https://user-images.githubusercontent.com/11146538/131616545-840d3284-9eda-481d-ba12-081e3dd38532.png">

https://github.com/1amageek/FirebaseSampleProject

**Comment by 1amageek** [comment]
<img width="923" alt="スクリーンショット 2021-09-01 14 37 16" src="https://user-images.githubusercontent.com/11146538/131617637-db7dd51d-aa9c-4ba9-a385-3bbcefdc5dca.png">

**Comment by 1amageek** [comment]
<img width="972" alt="スクリーンショット 2021-09-01 14 45 05" src="https://user-images.githubusercontent.com/11146538/131618362-5adef921-368e-46bc-a9a0-ec7c1bf916de.png">

**Comment by maksymmalyhin** [comment]
@1amageek Thank for providing the detailed examples, this is really helpful! Let us consider different options and come back with more detailed comments.

**Comment by morganchen12** [comment]
Hi @1amageek, I discussed this with the Firebase team and we haven't agreed upon a canonical way to generate effect-free structs for any Firestore document for use in app model layers. My conclusions on the solution you proposed:

>I believe the solution is to define lightweight DocumentReference, Timestamp and FieldValue structs in Firebase SDK iOS.

Timestamp and FieldValue, if written in Swift, should absolutely be `struct`s and not `class`es since they are effectively wrappers around specific proto values that the server will fill in later. It does not make sense in Objective-C to reference the same Timestamp object from multiple documents. If we could use Foundation's Swift/ObjC bridging, it would be great to have the Swift Timestamp struct be bridgeable to an ObjC Timestamp class.

DocumentReference is more complex because it defines the operations that can be performed on a document as well as the document location, and thus it's not trivially copyable or easily separated from the Firestore instance that it's created from.

For building a separate module to exclude Firebase from the SwiftUI preview, your best bet is to write your app model in structs that do not reference Firebase types (for instance, use Date instead of Timestamp), and then use test data in your SwiftUI preview to fill in standard types with special values like document IDs.

**Reply by 1amageek** [reply]
Hi @morganchen12 Thanks for the replies.

It may be a little different from what I envisioned, because it doesn't change the declaration of the Firebase SDK.

My suggestion was to provide it as a Swift feature like FilebaseFirestoreCombineSwift or FilebaseFunctionsCombineSwift.

`Swift4ServerTimestamp` is very close to this. [Swift4ServerTimestamp](https://github.com/firebase/firebase-ios-sdk/blob/bdac2b87f6d6e604deae914a970a75136b3a62b5/Firestore/Swift/Source/Codable/ServerTimestamp.swift#L129) is only used when using Codable. I think its existence is very nice. It makes sense to be able to declare it as a Model, because in the `Only UI Package` you don't need to use all the features of FieldValue and DocumentReference.

Wouldn't a lightweight Struct that is supposed to be Encoded and Decoded in the same way as Swift4ServerTimestamp be possible with `DocumentID` and `DocumentReference`?

For example, the DocumentReference `get` is never used in UI Package. It is expected to be used only for static data such as `documentID` and `path`.

**Reply by morganchen12** [reply]
That's definitely possible as an external library; I'm not sure if Firebase would have the bandwidth to build it into FirebaseFirestoreSwift. For value-type representations of DocumentReference, how would you initialize it without loading data from Firestore?

**Reply by 1amageek** [reply]
There is no need to integrate it into FirebaseFirestoreSwift.
The problem that needs to be solved is the slowdown of the Preview in SwiftUI development, due to the Firebase SDK build.

This means that it needs to exist as a separate package from FirebaseFirestoreSwift.

There is no need to initialize the DocumentReference. I think this level of expressiveness is sufficient. This structure should only need to revert to its original form when it goes through the FirestoreEncoder.

```swift
public struct DocumentID: Hashable, Codable {

    public var id: String

    public init(id: String) {
        self.id = id
    }

    public static var auto: Self {

        /// https://github.com/firebase/firebase-ios-sdk/blob/bdac2b87f6d6e604deae914a970a75136b3a62b5/Firestore/core/src/util/autoid.cc#L40
        let id = UUID().uuidString
        return .init(id: id)
    }
}

public struct DoucmentRefererence: Codable {

    public var path: String

    public var id: String { String(path.split(separator: "/").last!) }

    public init(path: String) {
        self.path = path
    }
}
```

**Reply by mortenbekditlevsen** [reply]
I think your question is really important since modularizing your app unlocks so many possibilities around testing, mocking, reusability, etc. 

My SwiftUI views only ever depends on two things: a DesignGuide framework containing ui components and a ViewModel which is a lightweight description of everything the view needs. The ViewModel has no further dependencies, and mainly consists of an ObservableObject class. 
The implementation that ties everything together is in a separate framework and subclasses the ViewModel and ties everything up to signals (combine or RxSwift or similar) from various services.

This way the ui is basically as lean as possible and ViewModels can be instantiated in previews in any state. 

Even the implementation of the logic driving the view models don't in fact depend on Firebase. That's abstracted away as a service that can get signals of model values from paths. Only at the top level is the actual Firebase implementation of that abstraction injected. 
That way all unit tests of these layers can run without Firebase too. 

I know this is likely too rigid for many people's taste, but you can loosen it up at many levels. For instance by allowing a framework containing domain models to be known by the ViewModel, and thus also by the ui. But this naturally requires you models to be quite simple and not have a bunch of further dependencies - and certainly not to Firebase, because then that's part of your previews again.

**Comment by nidegen** [comment]
I think one win would be to use `Identifiable` instead of `@DocumentID` for codable structs:)
